### DISCUSSION TITLE
Export 3D segments as ROI or Volumes, ImageJ compatibility

### DISCUSSION THREAD

**Discussion by DMAlessandro** [JacobBumgarner/VesselVio#54] [Ideas]
Hi, 
Great software, easy to use and implement in many pipelines. I'm adapting VessleVio to our light sheet microscopy data, however part of our analysis is in Imagej and it would be great to have access to the following features:

-Export segments as 3D ROIs or Volumes in ImageJ (specifically in our case we're looking for a way to filter the blood vessels by diameter and extract only capillaries 4-10microns)
-Implementation of VesselVio as a ImageJ/ FIJI suite plugin?
-Big Data handling for analysis and visualization of hierarchical data format (HDF) files (.h5) 

Thank you in advance for your consideration, 
Looking forward to the next updates

**Comment by JacobBumgarner** [comment]
Hey there @DMAlessandro , thanks for the kind notes. I'm glad to hear that VesselVio has been working for you.

In my free time I've been passively working on a large codebase rework, so I think that the first and third requests will likely be possible to implement as I get to the image processing rework. 

At the moment, I have no plans to port VesselVio to FIJI. This isn't because it wouldn't be useful, but mainly because I just don't have the time as I'm also trying to finish up my degree.

**Comment by JacobBumgarner** [comment]
So to dig down a bit more. 

> -Export segments as 3D ROIs or Volumes in ImageJ (specifically in our case we're looking for a way to filter the blood vessels by diameter and extract only capillaries 4-10microns)

So at the moment, you can export csv files for each dataset that contains the size of the vessels with their other features. You could use these CSV files to isolate the features of the small capillaries. Does this address what you're wanting? Otherwise, I'd be curious to hear about the desired use case for this. 

The difficulty with exporting sub-volumes is that VesselVio converts the voxel-based datasets into graph-based datasets and does centerline smoothing. This means that the location of the centerline points become floats rather than ints, so they wouldn't fit into an image grid.

In contrast, we could think about adding the option to create masks around the graphical vessels and converting the mask coordinates to ints.

**Reply by DMAlessandro** [reply]
Unfortunately I'm not familiar enough with image processing in python or with ITK, but I think obtaining masks the way you mention may be a way to go. 
Speaking about a use case scenario: while I'm using VesselVio for a general understanding of the whole mapped vasculature, I'd like to look at distance mapping of anoher fluorophore (channel 1) away from the imaged vessels (channel 2), specifically the capillaries (so a filtered subset of the vasculature). So, if there is a way to export the generated segment volumes/ surfaces/ masks in the same space and coordinates; more complex analysis could be made wth other fluorophores.
Again, I'm not familiar with image processing in Python, but I think once the surfaces are generated with VesselVio and identified (as they undergo analysis already), I imagine it could be possible to then export them (thinking of Imaris how the surfaces generated are superimposed on the original volume and exported).

**Comment by JacobBumgarner** [comment]
> -Big Data handling for analysis and visualization of hierarchical data format (HDF) files (.h5)

This is also a great idea; there have been a few other mentions of `.h5` compatibility in the past. The main reason I haven't added compatibility with this format is that I'm just not really familiar with it.

I believe that if a user were to load a `.h5` dataset, I'd also have to make them provide the key that points to the dataset in the file. Does this sound accurate?

**Reply by DMAlessandro** [reply]
I believe so, yes. If you were to load a hd5 or any pyramidal image in Bioformats (ImageJ) you can then choose the resolution you'd like to load in levels. If this were not possible, the conversion from .h5 to .tiff is easy enough, although metadata file may be lost in the process.

The main reason behind compatibility with .h5 files is that upon visualization of the results, a large (1Tb) file would not blow the whole memory.
